File: day16/day16.py
#!/usr/bin/env python3

import string
import sys

def dance(progs, move):
    if move[0] == "s":
        pos = len(progs) - int(move[1:])
        progs = progs[pos:] + progs[:pos]
    elif move[0] == "x":
        a, b = map(int, move[1:].split("/"))
        progs[a], progs[b] = progs[b], progs[a]
    elif move[0] == "p":
        a, b = map(progs.index, move[1:].split("/"))
        progs[a], progs[b] = progs[b], progs[a]
    return progs

def step(progs, txt):
    for move in txt.split(","):
        progs = dance(progs, move)
    return progs

def part1(txt, l):
    progs = list(string.ascii_lowercase[:l])
    return "".join(step(progs, txt))

def part2(txt, l):
    initial = list(string.ascii_lowercase[:l])
    tortoise = step(initial[:], txt)
    hare = step(step(initial[:], txt), txt)
    while tortoise != hare:
        tortoise = step(tortoise, txt)
        hare = step(hare, txt)
        hare = step(hare, txt)

    mu = 0
    tortoise = initial[:]
    while tortoise != hare:
        tortoise = step(tortoise, txt)
        hare = step(hare, txt)
        mu += 1

    lam = 1
    hare = step(tortoise[:], txt)
    while tortoise != hare:
        hare = step(hare, txt)
        lam += 1

    cycles = (1000000000 - mu) % lam + mu
    progs = initial
    for i in range(cycles):
        progs = step(progs, txt)

    return "".join(progs)

if __name__ == "__main__":
    if len(sys.argv) < 2:
        print("Usage: {} <filename>".format(sys.argv[0]), file=sys.stderr)
        sys.exit(1)

    try:
        with open(sys.argv[1], "rt") as f:
            txt = f.read().strip()
    except:
        print("Cannot open {}".format(sys.argv[1]), file=sys.stderr)
        sys.exit(1)

    #txt = "s1,x3/4,pe/b"
    print("Part1:", part1(txt, 16))
    print("Part2:", part2(txt, 16))

File: day16/day16.cpp
#include <fstream>
#include <iostream>

#include <fmt/format.h>

using namespace std;

struct Move
{
	enum {SPIN, EXCHANGE, PARTNER} type;
	int arg1;
	int arg2;

	void dance(string& progs) const
	{
		switch (type)
		{
		case SPIN: {
			for (int i = 0; i < arg1; i++)
			{
				int c = progs.back();
				progs.pop_back();
				progs.insert(progs.begin(), c);
			}
			break;
		}

		case EXCHANGE:{
			swap(progs[arg1], progs[arg2]);
			break;
		}

		case PARTNER: {
			size_t a = progs.find(arg1);
			size_t b = progs.find(arg2);
			swap(progs[a], progs[b]);
			break;
		}
		}
	}
};

istream& operator>>(istream& in, Move& m)
{
	switch(in.get())
	{
	case 's':
		m.type = Move::SPIN;
		in >> m.arg1;
		break;
	case 'x':
		m.type = Move::EXCHANGE;
		in >> m.arg1;
		in.get();
		in >> m.arg2;
		break;
	case 'p':
		m.type = Move::PARTNER;
		m.arg1 = in.get();
		in.get();
		m.arg2 = in.get();
		break;
	default:
		in.setstate(ios::failbit);
		break;
	}
	return in;
}

static void step(const vector<Move>& moves, string& progs)
{
	for (auto& m: moves)
	{
		m.dance(progs);
	}
}

static string part1(const vector<Move>& moves)
{
	string initial;
	for (int i = 0; i < 16; i++)
	{
		initial.push_back('a' + i);
	}
	step(moves, initial);
	return initial;
}

static string part2(const vector<Move>& moves)
{
	string initial;
	for (int i = 0; i < 16; i++)
	{
		initial.push_back('a' + i);
	}

	// floyd algorithm
	string tortoise(initial);
	string hare(initial);
	do
	{
		step(moves, tortoise);
		step(moves, hare);
		step(moves, hare);
	} while (tortoise != hare);

	// find mu
	size_t mu = 0;
	tortoise = initial;
	while (tortoise != hare)
	{
		step(moves, tortoise);
		step(moves, hare);
		mu++;
	}

	// find lambda
	size_t lam = 0;
	hare = tortoise;
	do
	{
		step(moves, hare);
		lam++;
	} while (tortoise != hare);

	size_t cycles = (1000000000ULL - mu) % lam + mu;
	while (cycles-->0)
	{
		step(moves, initial);
	}
	return initial;
}


int main(int argc, char *argv[])
{
	if (argc < 2)
	{
		cerr << fmt::format("Usage: {} <filename>", argv[0]) << endl;
		return 1;
	}

	ifstream input(argv[1]);
	if (!input)
	{
		cerr << fmt::format("Cannot open {}", argv[1]) << endl;
		return 1;
	}
	vector<Move> moves;
	Move m;
	while (input>>m)
	{
		input.get();
		moves.emplace_back(move(m));
	}
	input.close();

	cout << fmt::format("Part1: {}", part1(moves)) << endl
	     << fmt::format("Part2: {}", part2(moves)) << endl;
	return 0;
}
